### PR TITLE
JDK-8197536: TableView, ListView: unexpected scrolling behaviour on up/down keys

### PR DESCRIPTION
**Pull request by tsayao**
https://bugs.openjdk.java.net/browse/JDK-8197536

As pointed out in the bug it was introduced when VirtualFlow moved to the public scope. There was a special case for scrolling up/down one cell that was removed.

This fix restores the special case, that is more precise than the 0 to 1 relative position on the virtual flow.

The 0 to 1 (0% to 100%) position of the vertical scroll is not cell height precise, so the special case is needed for "move one cell up" ou "move one cell down". (specially with long lists).

Is a test case doable?

### CODE REVIEW

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java [hunk 1]
```diff
@@ -1458,12 +1458,44 @@ public void scrollTo(int index) {
         if (cell != null) {
             scrollTo(cell);
         } else {
+            //see JDK-8197536
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: normally we put a space after the `//`

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java [hunk 2]
```diff
@@ -1458,12 +1458,44 @@ public void scrollTo(int index) {
...
+            if(tryScrollOneUpOrDown(index, true))
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: Needs a space after `if` before the `(`

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java [hunk 3]
```diff
@@ -1458,12 +1458,44 @@ public void scrollTo(int index) {
...
+        //if going down, cell diff is -1, because it will get the target cell index and check if previous
+        //cell is visible to base the position
+        int indexDiff = (down) ? -1 : 1;
+
+        T next = getVisibleCell(targetIndex + indexDiff);
+        if (next != null) {
+            T cell = getAvailableCell(targetIndex);
+            setCellIndex(cell, targetIndex);
+            resizeCell(cell); // resize must be after config
+            if(down) {
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: space after `if`

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java [hunk 4]
```diff
@@ -1458,12 +1458,44 @@ public void scrollTo(int index) {
...
             addAllToPile();
             requestLayout();
         }
     }
 
+    //will return true if scroll is successful
+    private boolean tryScrollOneUpOrDown(int targetIndex, boolean down) {
+        //if going down, cell diff is -1, because it will get the target cell index and check if previous
+        //cell is visible to base the position
+        int indexDiff = (down) ? -1 : 1;
```
Comment (brcolow) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: Nit-pick: I would remove the unnecessary parentheses around `(down)`.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 1 above)
Comment (tsayao) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: done!

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 3 above)
Comment (tsayao) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: ok!

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 4 above)
Comment (tsayao) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: done!

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 2 above)
Comment (tsayao) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: done!

**Comment by tsayao** [comment]
This test below shows that horizontal scroll behavior is not altered.

However, if the setCellSelectionEnabled(true) line is commented the horizontal scroll behavior is weird because it's per cell even if the selection is per row (with or without the changes).

Column generation based on:
https://stackoverflow.com/questions/37559584/how-to-add-dynamic-columns-and-rows-to-tableview-in-java-fxml

```
import javafx.application.Application;
import javafx.beans.property.ReadOnlyObjectWrapper;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.TableColumn;
import javafx.scene.control.TableView;
import javafx.stage.Stage;

import java.util.ArrayList;
import java.util.List;

public class TableViewScroll extends Application {
    private static final int N_COLS = 50;
    private static final int N_ROWS = 10000;

    private Parent getContent() {

        TestDataGenerator dataGenerator = new TestDataGenerator();

        TableView<ObservableList<String>> tableView = new TableView<>();

        // add columns
        List<String> columnNames = dataGenerator.getNext();
        for (int i = 0; i < columnNames.size(); i++) {
            final int finalIdx = i;
            TableColumn<ObservableList<String>, String> column = new TableColumn<>(
                    columnNames.get(i)
            );
            column.setCellValueFactory(param ->
                    new ReadOnlyObjectWrapper<>(param.getValue().get(finalIdx))
            );
            tableView.getColumns().add(column);
        }

        // add data
        for (int i = 0; i < N_ROWS; i++) {
            tableView.getItems().add(
                    FXCollections.observableArrayList(
                            dataGenerator.getNext()
                    )
            );
        }

        tableView.setPrefHeight(200);

        tableView.getSelectionModel().setCellSelectionEnabled(true);
        return tableView;
    }

    @Override
    public void start(Stage primaryStage) throws Exception {
        primaryStage.setScene(new Scene(getContent(), 800, 400));
        primaryStage.show();

    }

    public static void main(String[] args) {
        launch(args);
    }

    private static class TestDataGenerator {
        private static final String[] LOREM = "The quick brown fox jumped over the lazy dog and got run over by crazy maniac driver.".split(" ");

        private int curWord = 0;

        List<String> getNext() {
            List<String> words = new ArrayList<>();

            for (int i = 0; i < TableViewScroll.N_COLS; i++) {
                if (curWord == Integer.MAX_VALUE) {
                    curWord = 0;
                }

                words.add(LOREM[curWord % LOREM.length]);
                curWord++;
            }

            return words;
        }
    }
}
```

**Comment by kevinrushforth** [comment]
> This test below shows that horizontal scroll behavior is not altered.

I was actually asking about a horizontal `ListView`, not `TableView`. The test program attached to [JDK-8197536](https://bugs.openjdk.java.net/browse/JDK-8197536) also has a `ListView`.

If I add the following call to that program:

```
    list.setOrientation(Orientation.HORIZONTAL);
```

it (mis)behaves in the same way as the vertical `ListVIew` does without your patch. With your patch, some things get better, but not consistently. There are some cases now where only a partial list item is visible. This might be due to the fact that each list item has a different width in the horizontal case (as opposed to the vertical list, which is the same). This will need to be evaluated, since I believe it is possible for rows of a `ListView` in vertical mode to be of different heights. Similarly, a `TableView` might have rows of differing height.

**Comment by tsayao** [comment]
New test app below.

Still researching a solution for the horizontal list.

```
import javafx.application.Application;
import javafx.beans.property.ReadOnlyObjectWrapper;
import javafx.collections.FXCollections;
import javafx.collections.ObservableList;
import javafx.geometry.Orientation;
import javafx.geometry.Pos;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.*;
import javafx.scene.layout.BorderPane;
import javafx.scene.layout.Pane;
import javafx.stage.Stage;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class TableViewScroll extends Application {
    private static final int N_ROWS = 5000;

    private Parent getContent() {
        TestDataGenerator dataGenerator = new TestDataGenerator(100);
        TableView<ObservableList<String>> tableView = new TableView<>();

        // add columns
        List<String> columnNames = dataGenerator.getNext();
        for (int i = 0; i < columnNames.size(); i++) {
            final int finalIdx = i;
            TableColumn<ObservableList<String>, String> column = new TableColumn<>(columnNames.get(i));
            column.setCellValueFactory(param -> new ReadOnlyObjectWrapper<>(param.getValue().get(finalIdx)));

            tableView.getColumns().add(column);
        }

        // add data
        for (int i = 0; i < N_ROWS; i++) {
            tableView.getItems().add(FXCollections.observableArrayList(dataGenerator.getNext()));
        }

        BorderPane pane = new BorderPane();

        pane.setCenter(tableView);

        TestDataGenerator dataGenerator2 = new TestDataGenerator(1000);
        ListView<String> list = new ListView<>();
        list.setItems(FXCollections.observableArrayList(dataGenerator2.getNext()));
        pane.setRight(list);

        ListView<String> hList = new ListView<>();
        hList.setOrientation(Orientation.HORIZONTAL);
        hList.setItems(FXCollections.observableArrayList(dataGenerator2.getNext()));
        hList.setPrefHeight(200);
        pane.setBottom(hList);

        pane.setPrefSize(800, 600);

        return pane;
    }

    @Override
    public void start(Stage primaryStage) throws Exception {
        primaryStage.setScene(new Scene(getContent()));
        primaryStage.show();
    }

    public static void main(String[] args) {
        launch(args);
    }

    private static class TestDataGenerator {
        private static final String[] LOREM = "The quick brown fox jumped over the lazy dog and got run over by crazy maniac driver.".split(" ");
        private int curWord = 0;
        private final int cols;

        TestDataGenerator(int cols) {
            this.cols = cols;
        }

        List<String> getNext() {
            List<String> words = new ArrayList<>();

            for (int i = 0; i < cols; i++) {
                if (curWord == Integer.MAX_VALUE) {
                    curWord = 0;
                }

                words.add(LOREM[curWord % LOREM.length]);
                curWord++;
            }

            return words;
        }
    }
}
```

**Comment by tsayao** [comment]
Kevin,

I have improved it to work better with horizontal orientation.

The comment above has an updated test code.

Will try to provide a test case (not sure how yet).

**Comment by tsayao** [comment]
Test case provided.

**Comment by tsayao** [comment]
This PR is ready.

**Comment by kevinrushforth** [comment]
> This PR is ready.

Thanks.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java [hunk 5]
```diff
@@ -1458,12 +1458,45 @@ public void scrollTo(int index) {
         if (cell != null) {
             scrollTo(cell);
         } else {
+            // see JDK-8197536
+            if (tryScrollOneCell(index, true))
+                return;
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: Please enclose this in curly braces; our Java code conventions do not allow for single-statement targets of a conditional expression (unless on the same line as the conditional, which isn't desirable here).

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java [hunk 6]
```diff
@@ -1458,12 +1458,45 @@ public void scrollTo(int index) {
...
+            else if (tryScrollOneCell(index, false))
+                return;
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: Please enclose this in curly braces.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/VirtualFlowTest.java [hunk 7]
```diff
@@ -1126,6 +1126,44 @@ public void test_RT39568() {
...
+        double viewportLength = flow.getViewportLength();
+
+        assertEquals("Last cell must end on viewport size", viewportLength, (cellPosition + cellLength), 0.1);
+    }
+
+    @Test
+    // see JDK-8197536
+    public void testScrollOneCellHorizontal() {
+        flow.setVertical(false);
+        testScrollOneCell();
```
Comment (kevinrushforth) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/VirtualFlowTest.java]: Minor: Normally we wouldn't call one test method from another. Rather we typically create a utility method that both test methods would call.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 5 above)
Comment (tsayao) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: Done. Thanks.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java (hunk 6 above)
Comment (tsayao) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: Ok

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/VirtualFlowTest.java (hunk 7 above)
Comment (tsayao) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/skin/VirtualFlowTest.java]: Moved to a method as requested. Thanks.

File: modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java [hunk 8]
```diff
@@ -1458,12 +1458,46 @@ public void scrollTo(int index) {
...
             addAllToPile();
             requestLayout();
         }
     }
 
+    // will return true if scroll is successful
+    private boolean tryScrollOneCell(int targetIndex, boolean downOrRight) {
+        // if going down, cell diff is -1, because it will get the target cell index and check if previous
+        // cell is visible to base the position
+        int indexDiff = downOrRight ? -1 : 1;
```
Comment (aghaisas) [inline modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: Remove space before '?'

Comment (kevinrushforth) [reply modules/javafx.controls/src/main/java/javafx/scene/control/skin/VirtualFlow.java]: No, the space is consistent with our coding style.

**Pull request merged**